`wxListCtrl` fixes in wxMSW

vadz opened the pull request:
This is #25950 with minor changes and split into several commits.

Crementif committed:
Draw wxListCtrl subitems in visual, not logical, order

This may avoid issues due to overflowing into the next column.

Crementif committed:
Make wxListCtrl dark mode closer to native, white mode rendering

Avoid problems due to overflowing into the next column by checking that
there is enough space available before drawing.

Adjust margins and padding to try to replicate what Windows uses
natively.

Crementif committed:
Fix wxListCtrl::GetSubItemRect(n, 0) when columns are reordered

Sum up all the preceding column widths to find the starting position of
the first column, which may not be 0 if the columns have been reordered.

Crementif committed:
Erase the whole list item row before drawing sub-items

Ensure hover colour doesn't remain anywhere, e.g. it doesn't show
through check boxes or (transparent) icons.

MaartenBent commented on src/msw/listctrl.cpp:
```diff
@@ -3218,32 +3288,45 @@ void HandleItemPaint(wxListCtrl* listctrl, LPNMLVCUSTOMDRAW pLVCD, HFONT hfont)
+        if ( !listctrl->GetSubItemRect(item, col, rectSubItem, wxLIST_RECT_BOUNDS) )
+            continue;
+        RECT rcSubItem;
+        wxCopyRectToRECT(rectSubItem, rcSubItem);
```
Can get the `RECT` directly like:
```suggestion
        RECT rcSubItem;
        if ( !wxGetListCtrlSubItemRect(nmcd.hdr.hwndFrom, item, col, LVIR_BOUNDS, rcSubItem) )
            continue;
```

MaartenBent replied:
> ListCtrl::GetSubItemRect() does additional math to normalize the output of wxGetListCtrlSubItemRect, since its fully incorrect.

Sorry missed your message. Indeed, I also noticed wrong x-offset when reordering columns.

MaartenBent commented on src/msw/listctrl.cpp:
```diff
@@ -3218,32 +3288,45 @@ void HandleItemPaint(wxListCtrl* listctrl, LPNMLVCUSTOMDRAW pLVCD, HFONT hfont)
+        RECT rcSubItem;
+        wxCopyRectToRECT(rectSubItem, rcSubItem);
+
+        HandleSubItemPrepaint(listctrl, pLVCD, hfont, rcSubItem, pLVCD->clrTextBk);
```
You already pass `pLVCD` to the function, so `pLVCD->clrTextBk` is redundant.

vadz commented:
@Crementif If you don't have any objections, I'll merge this soon.

MaartenBent committed:
Improve right-aligned text positioning in wxListCtrl dark mode

Handle it correctly, i.e. symmetrically to the left-aligned text and not
in the same way as it.

Crementif commented on src/msw/listctrl.cpp:
```diff
@@ -3050,41 +3074,52 @@ bool HandleSubItemPrepaint(wxListCtrl* listctrl, LPNMLVCUSTOMDRAW pLVCD, HFONT h
+    const int availableWidth = wxMax(rc.right - rc.left, 0);
 
-    if ( !col && listctrl->HasCheckBoxes() )
+    if ( !col && listctrl->HasCheckBoxes() && availableWidth > 0 )
```
I think this is a mistake. Here, with the right column width, the availableWidth could end up being 1-2. But the gap between the left side and the checkbox is also 2, but we still do a call to ImageList_DrawEx.

Crementif replied:
I meant the opposite, sorry. Ultimately it should follow the native white mode behavior, which is that it always cuts the content off correctly. The bug still stands however, at least from what I can see.

vadz replied:
I might be missing something (it's getting late here...), but I don't see any obvious difference in behaviour between light and dark modes with the current version, the checkbox does get clipped when the column width becomes very small. Maybe it's because you've also changed code to overdraw the columns by the following ones?

Anyhow, what about this diff?

```diff
diff --git a/src/msw/listctrl.cpp b/src/msw/listctrl.cpp
index e77b4450d5..9a219ba11c 100644
--- a/src/msw/listctrl.cpp
+++ b/src/msw/listctrl.cpp
@@ -3079,8 +3079,9 @@ HandleSubItemPrepaint(wxListCtrl* listctrl,
     rc.right -= PADDING_RIGHT_SIDE;
 
     const int availableWidth = wxMax(rc.right - rc.left, 0);
+    const int availableForCheckbox = availableWidth - GAP_AFTER_CHECKBOX;
 
-    if ( !col && listctrl->HasCheckBoxes() && availableWidth > 0 )
+    if ( !col && listctrl->HasCheckBoxes() && availableForCheckbox > 0 )
     {
         const HIMAGELIST himl = ListView_GetImageList(hwndList, LVSIL_STATE);
 
@@ -3095,7 +3096,7 @@ HandleSubItemPrepaint(wxListCtrl* listctrl,
             int cbY = rc.top + ((rc.bottom - rc.top) / 2 - cbHeight / 2);
 
             // prevent drawing checkbox farther than the column width
-            cbWidth = wxClip(cbWidth, 0, availableWidth);
+            cbWidth = wxClip(cbWidth, 0, availableForCheckbox);
 
             // When using style flag ILD_SELECTED or ILD_FOCUS, the checkboxes
             // for selected items are drawn with a blue background, which we want to avoid.
```

I don't see any visual differences but it does seem to be more logical to do it like this.

Crementif commented on src/msw/listctrl.cpp:
```diff
@@ -3050,41 +3074,52 @@ bool HandleSubItemPrepaint(wxListCtrl* listctrl, LPNMLVCUSTOMDRAW pLVCD, HFONT h
+            int cbY = rc.top + ((rc.bottom - rc.top) / 2 - cbHeight / 2);
+
+            // prevent drawing checkbox farther than the column width
+            cbWidth = wxClip(cbWidth, 0, availableWidth);
```
availableWidth isn't the current remaining space that's available in the column, since we've added GAP_BEFORE_CHECKBOX to it already. The checkbox is now able to overflow the next column by 4 pixels.

vadz replied:
Oops, thanks, I've indeed missed it. But how to fix this depends on the fix for the comment above, i.e. I would just replace `availableWidth > 0` with `availableWidth > GAP_BEFORE_CHECKBOX` and use `availableWidth - GAP_BEFORE_CHECKBOX` here, but not if we remove that check completely...

vadz replied:
What about the latest version? I've removed `availableWidth` to get a bit closer back to the original version, but I've also added returns to make sure we don't try drawing anything which is certain not to fit.

vadz commented on src/msw/listctrl.cpp:
```diff
@@ -3137,6 +3203,7 @@ bool HandleSubItemPrepaint(wxListCtrl* listctrl, LPNMLVCUSTOMDRAW pLVCD, HFONT h
         {
             case LVCFMT_LEFT:
                 fmt |= DT_LEFT;
+                rc.left += PADDING_FOR_TEXT;
```
I think it's relatively clear as it's only used in this one place. I did remove

```cpp
// text is drawn 1 unit away from the background rectangle
```

comment intentionally because it seemed to just repeat what the code was quite clearly doing.

OTOH looking at this code again, I think we should apply padding (at both sides) for centered text too, shouldn't we?

Crementif replied:
Hard to say without testing, but maybe?

vadz committed:
Make HandleSubItemPrepaint() void

The return value of this function was never used anyhow, so don't bother
returning anything from it.

vadz committed:
Skip drawing the rest of the item if there is not enough space

Don't bother continuing with drawing the image if there is not enough
space for the checkbox or drawing the text if there is not enough space
for the image.
